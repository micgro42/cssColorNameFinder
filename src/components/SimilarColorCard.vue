<template>
  <div class="similarColorCard">
    <div class="similarColorCard__colorArea">
      <div
        class="similarColorCard__colorHalf"
        :style="'background-color: ' + colorName"
      />
      <div
        class="similarColorCard__colorHalf"
        :style="'background-color: ' + originalHex"
      />
    </div>
    <ColorNames
      :color-name="colorName"
      :color-hex="similarHex"
      :original-hex="originalHex"
    />
    <div class="similarColorCard__comparison">
      <ColorDifferenceDetail
        section-name="Lightness"
        :similar-value="round(similarColor.lightness, 2).toString()"
        :original-value="round(originalColor.lightness, 2).toString()"
      >
        <p v-if="lightnessDiff > 0">
          <i>{{ colorName }}</i> is
          <b>{{ lightnessDiff }}&nbsp;points lighter</b>
          than <i>{{ originalHex }}</i
          >.
        </p>
        <p v-else-if="lightnessDiff < 0">
          <i>{{ colorName }}</i> is
          <b>{{ Math.abs(lightnessDiff) }}&nbsp;points darker</b>
          than <i>{{ originalHex }}</i
          >.
        </p>
        <p v-else>
          <i>{{ colorName }}</i> is
          <b>exactly as light</b>
          as <i>{{ originalHex }}</i
          >.
        </p>
      </ColorDifferenceDetail>
      <ColorDifferenceDetail
        section-name="Chroma"
        :similar-value="`${round(similarColor.chroma, 2)}`"
        :original-value="`${round(originalColor.chroma, 2)}`"
      >
        <p v-if="chromaDiff > 0">
          <i>{{ colorName }}</i> is
          <b>{{ chromaDiff }}&nbsp;points more intense</b>
          than <i>{{ originalHex }}</i
          >.
        </p>
        <p v-else-if="chromaDiff < 0">
          <i>{{ colorName }}</i> is
          <b>{{ Math.abs(chromaDiff) }}&nbsp;points less intense</b>
          than <i>{{ originalHex }}</i
          >.
        </p>
        <p v-else>
          <i>{{ colorName }}</i> is
          <b>exactly as intense</b>
          as <i>{{ originalHex }}</i
          >.
        </p>
      </ColorDifferenceDetail>
      <ColorDifferenceDetail
        v-if="!isNaN(hueDiff)"
        section-name="LCH Hue"
        :similar-value="`${round(similarColor.hue, 2)}°`"
        :original-value="`${round(originalColor.hue, 2)}°`"
      >
        <p>
          <i>{{ colorName }}</i> differs by
          <b>{{ hueDiff }}&nbsp;degrees</b>
          from <i>{{ originalHex }}</i
          >.
        </p>
      </ColorDifferenceDetail>
    </div>
  </div>
</template>

<script lang="ts">
import { NamedColorData, ColorData } from '@/color/similarNamedColors';
import ColorDifferenceDetail from '@/components/ColorDifferenceDetail.vue';
import ColorNames from '@/components/ColorNames.vue';
import { defineComponent, PropType } from 'vue';

export default defineComponent({
  components: { ColorDifferenceDetail, ColorNames },
  props: {
    colorName: {
      type: String,
      required: true,
    },
    similarColor: {
      type: Object as PropType<NamedColorData>,
      required: true,
    },
    originalColor: {
      type: Object as PropType<ColorData>,
      required: true,
    },
  },
  computed: {
    originalHex(): string {
      return (
        '#' +
        this.originalColor.srgb
          .map((fraction: number) => (fraction * 255).toString(16).padStart(2, '0'))
          .join('')
      );
    },
    similarHex(): string {
      return (
        '#' +
        this.similarColor.srgb
          .map((fraction: number) => (fraction * 255).toString(16).padStart(2, '0'))
          .join('')
      );
    },
    lightnessDiff(): number {
      const lightnessDiff = this.similarColor.lightness - this.originalColor.lightness;
      return this.round(lightnessDiff, 2);
    },
    chromaDiff(): number {
      const chromaDiff = this.similarColor.chroma - this.originalColor.chroma;
      return this.round(chromaDiff, 2);
    },
    hueDiff(): number {
      const normalizedSimilarHue =
        this.similarColor.hue > 180 ? this.similarColor.hue - 360 : this.similarColor.hue;
      const normalizedOriginalHue =
        this.originalColor.hue > 180 ? this.originalColor.hue - 360 : this.originalColor.hue;
      const diff = normalizedSimilarHue - normalizedOriginalHue;
      return this.round(Math.abs(diff), 2);
    },
  },
  methods: {
    round(number: number, decimals = 0): number {
      return (
        Math.round((number + Number.EPSILON) * Math.pow(10, decimals)) / Math.pow(10, decimals)
      );
    },
  },
});
</script>

<style lang="scss">
.similarColorCard {
  margin-top: 1rem;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  background-color: white;
}

.similarColorCard__colorArea {
  min-height: 10vh;
  display: flex;
}

.similarColorCard__colorHalf {
  width: 50%;
}
</style>
