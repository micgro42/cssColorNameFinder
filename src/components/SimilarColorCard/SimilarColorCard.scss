.similarColorCard {
  margin-top: 1rem;
  padding: 1rem;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  background-color: white;
}

.similarColorCard__header {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
}

.similarColorCard__colorArea {
  min-height: 10vh;
  border: 1px solid grey;
}
