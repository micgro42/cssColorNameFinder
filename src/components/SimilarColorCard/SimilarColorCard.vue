<template>
    <div class="similarColorCard">
        <div class="similarColorCard__header">
            <h3>{{ colorName }}</h3> <p>Distance: {{ roundedDistance }}</p>
        </div>
        <div
                class="similarColorCard__colorArea"
                :style="'background-color: ' + colorName"
        ></div>
    </div>
</template>

<script lang="ts">

interface SimilarColorCardThis {
    distance: number;
    colorName: string;
    roundedDistance: () => number
}

export default {
    props: {
        colorName: {
            type: String,
            required: true,
        },
        distance: {
            type: Number,
            required: true,
        },
    },
    computed: {
        roundedDistance() {
            return Math.round((this as SimilarColorCardThis).distance);
        }
    }
};
</script>

<style scoped lang="scss">
    .similarColorCard {
        margin-top: 1rem;
        padding: 1rem;
        box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
        background-color: white;
    }

    .similarColorCard__header {
        display: flex;
        justify-content: space-between;
        align-items: baseline;
    }

    .similarColorCard__colorArea {
        min-height: 10vh;
        border: 1px solid grey;
    }
</style>
