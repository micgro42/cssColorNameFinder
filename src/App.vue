<template>
  <div class="app">
    <h1>CSS Color Name Finder</h1>
    <IntroCard v-if="similarColors.length === 0" />
    <label>
      Pick a color:
      <input
        type="color"
        @change="changeColor"
      />
    </label>
    <div v-if="originalColor && similarColors.length">
      <h2>Colors close to {{ name }}:</h2>
      <SimilarColorCard
        v-for="color in similarColors"
        :key="color.name"
        :color-name="color.name"
        :similar-color="color"
        :original-color="originalColor"
      />
    </div>
  </div>
</template>

<script lang="ts">
import getCloseNamedColors, {
  ColorData,
  getColorData,
  NamedColorData,
} from '@/color/similarNamedColors';
import IntroCard from '@/components/IntroCard.vue';
import SimilarColorCard from '@/components/SimilarColorCard.vue';
import { defineComponent } from 'vue';

export default defineComponent({
  components: {
    IntroCard,
    SimilarColorCard,
  },
  data() {
    return {
      name: '',
    };
  },
  computed: {
    similarColors(): NamedColorData[] {
      if (!this.name) {
        return [];
      }
      const closeNamedColors = getCloseNamedColors(this.name);
      return closeNamedColors.slice(0, 3);
    },
    originalColor(): ColorData | null {
      if (!this.name) {
        return null;
      }
      return getColorData(this.name);
    },
  },
  methods: {
    changeColor(event: Event): void {
      this.name = (event.target as HTMLInputElement).value;
    },
  },
});
</script>
