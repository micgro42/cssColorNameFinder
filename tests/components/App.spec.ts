import App from '@/App.vue';
import SimilarColorCard from '@/components/SimilarColorCard/SimilarColorCard.vue';
import { shallowMount } from '@vue/test-utils';

describe('App', () => {
    it('shows similar color if color was selected', async () => {
        const wrapper = shallowMount(App);
        wrapper.setData({ name: '#abcdef' });
        await wrapper.vm.$nextTick();
        const similarColorCards = wrapper.findAllComponents(SimilarColorCard);
        expect(similarColorCards.length).toBe(3);
        expect(similarColorCards[0]?.props()).toStrictEqual({ colorName: 'lightblue', distance: 14.352700094407323 });
        expect(similarColorCards[1]?.props()).toStrictEqual({
            colorName: 'lightsteelblue',
            distance: 19.87460691435179,
        });
        expect(similarColorCards[2]?.props()).toStrictEqual({ colorName: 'powderblue', distance: 21.61018278497431 });
    });
});
