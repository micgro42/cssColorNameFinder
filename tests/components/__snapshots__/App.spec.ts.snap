// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shows similar colors if color was selected2 1`] = `
<div
  class="app"
>
  <h1>
    CSS Color Name Finder
  </h1>
  <!--v-if-->
  <label>
     Pick a color: 
    <input
      type="color"
    />
  </label>
  <div>
    <h2>
      Colors close to #abcdef:
    </h2>
    
    <div
      class="similarColorCard"
    >
      <div
        class="similarColorCard__colorArea"
      >
        <div
          class="similarColorCard__colorHalf"
          style="background-color: lightsteelblue;"
        />
        <div
          class="similarColorCard__colorHalf"
          style="background-color: rgb(171, 205, 239);"
        />
      </div>
      <div
        class="ColorNames"
      >
        <div
          class="ColorNames__names"
        >
          <!-- width 50%! -->
          <h3>
            lightsteelblue
          </h3>
          <h3>
            your color:
          </h3>
        </div>
        <div
          class="ColorNames__values"
        >
          <span>
            #b0c4de
          </span>
          <span>
            #abcdef
          </span>
        </div>
      </div>
      <div
        class="similarColorCard__comparison"
      >
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            Lightness
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              78.26
            </span>
            <span>
              80.77
            </span>
          </div>
          
          <p>
            <i>
              lightsteelblue
            </i>
             is 
            <b>
              2.51 points darker
            </b>
             than 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            Chroma
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              15.7
            </span>
            <span>
              21.62
            </span>
          </div>
          
          <p>
            <i>
              lightsteelblue
            </i>
             is 
            <b>
              5.92 points less intense
            </b>
             than 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            LCH Hue
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              259.35°
            </span>
            <span>
              254.01°
            </span>
          </div>
          
          <p>
            <i>
              lightsteelblue
            </i>
             differs by 
            <b>
              5.34 degrees
            </b>
             from 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
      </div>
    </div>
    <div
      class="similarColorCard"
    >
      <div
        class="similarColorCard__colorArea"
      >
        <div
          class="similarColorCard__colorHalf"
          style="background-color: lightskyblue;"
        />
        <div
          class="similarColorCard__colorHalf"
          style="background-color: rgb(171, 205, 239);"
        />
      </div>
      <div
        class="ColorNames"
      >
        <div
          class="ColorNames__names"
        >
          <!-- width 50%! -->
          <h3>
            lightskyblue
          </h3>
          <h3>
            your color:
          </h3>
        </div>
        <div
          class="ColorNames__values"
        >
          <span>
            #87cefa
          </span>
          <span>
            #abcdef
          </span>
        </div>
      </div>
      <div
        class="similarColorCard__comparison"
      >
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            Lightness
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              79.29
            </span>
            <span>
              80.77
            </span>
          </div>
          
          <p>
            <i>
              lightskyblue
            </i>
             is 
            <b>
              1.48 points darker
            </b>
             than 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            Chroma
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              32.48
            </span>
            <span>
              21.62
            </span>
          </div>
          
          <p>
            <i>
              lightskyblue
            </i>
             is 
            <b>
              10.86 points more intense
            </b>
             than 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            LCH Hue
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              243.62°
            </span>
            <span>
              254.01°
            </span>
          </div>
          
          <p>
            <i>
              lightskyblue
            </i>
             differs by 
            <b>
              10.39 degrees
            </b>
             from 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
      </div>
    </div>
    <div
      class="similarColorCard"
    >
      <div
        class="similarColorCard__colorArea"
      >
        <div
          class="similarColorCard__colorHalf"
          style="background-color: skyblue;"
        />
        <div
          class="similarColorCard__colorHalf"
          style="background-color: rgb(171, 205, 239);"
        />
      </div>
      <div
        class="ColorNames"
      >
        <div
          class="ColorNames__names"
        >
          <!-- width 50%! -->
          <h3>
            skyblue
          </h3>
          <h3>
            your color:
          </h3>
        </div>
        <div
          class="ColorNames__values"
        >
          <span>
            #87ceeb
          </span>
          <span>
            #abcdef
          </span>
        </div>
      </div>
      <div
        class="similarColorCard__comparison"
      >
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            Lightness
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              78.85
            </span>
            <span>
              80.77
            </span>
          </div>
          
          <p>
            <i>
              skyblue
            </i>
             is 
            <b>
              1.92 points darker
            </b>
             than 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            Chroma
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              27.97
            </span>
            <span>
              21.62
            </span>
          </div>
          
          <p>
            <i>
              skyblue
            </i>
             is 
            <b>
              6.35 points more intense
            </b>
             than 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
        <div
          class="ColorDifferenceDetail"
        >
          <h4
            class="ColorDifferenceDetail__headline"
          >
            LCH Hue
          </h4>
          <div
            class="ColorDifferenceDetail__values"
          >
            <span>
              231.17°
            </span>
            <span>
              254.01°
            </span>
          </div>
          
          <p>
            <i>
              skyblue
            </i>
             differs by 
            <b>
              22.84 degrees
            </b>
             from 
            <i>
              #abcdef
            </i>
            . 
          </p>
          
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
